client/src/ui/molecules/language-menu/index.tsx

```diff
@@ -19,13 +21,36 @@ export function LanguageMenu({
   native: string;
 }) {
   const ga = useGA();
-  const { pathname } = useLocation();
+  const { hash, pathname } = useLocation();
   const navigate = useNavigate();
   const [preferredLocale, setPreferredLocale] = React.useState(locale);
 
-  function translateURL(destinationLocale: string) {
-    return pathname.replace(`/${locale}/`, `/${destinationLocale}/`);
-  }
+  // This effect makes you automatically navigate to the locale your cookie
+  // prefers if the current page's locale isn't what you prefer and the
+  // locale you prefer is one of the valid translations.
+  React.useEffect(() => {
+    const cookieLocale = getPreferredCookieLocale(document);
+    if (
+      locale &&
+      cookieLocale &&
+      locale.toLowerCase() !== cookieLocale.toLowerCase() &&
+      // If the URL is something like `#localeOverride` we omit this
+      // automatic "redirect" because the user has most likely clicked
+      // a link that means that want to "peek" at a locale that is
+      // different from what their cookie prefers.
+      !hash.toLowerCase().includes(LOCALE_OVERRIDE_HASH.toLowerCase()) &&
+      translations
+        .map((t) => t.locale.toLowerCase())
+        .includes(cookieLocale.toLowerCase())
+    ) {
+      const newPathname = translateURL(pathname, locale, cookieLocale);
+      // Just to be absolutely paranoidly certain it's not going to redirect
+      // to the URL you're already don, we're doing this extra check.
+      if (newPathname !== pathname) {
+        navigate(newPathname);
+      }
+    }
+  }, [locale, hash, pathname, navigate, translations]);
```

Comment:
```suggestion
  React.useEffect(() => {
    const cookieLocale = getPreferredCookieLocale(document);
    const hasLocaleOverrideHash = hash
      .toLowerCase()
      .includes(LOCALE_OVERRIDE_HASH.toLowerCase());

    if (
      !preventLocaleOverride &&
      cookieLocale &&
      locale?.toLowerCase() !== cookieLocale.toLowerCase() &&
      // If the URL is something like `#localeOverride` we omit this
      // automatic "redirect" because the user has most likely clicked
      // a link that means that want to "peek" at a locale that is
      // different from what their cookie prefers.
      !hasLocaleOverrideHash &&
      translations
        .map((t) => t.locale.toLowerCase())
        .includes(cookieLocale.toLowerCase())
    ) {
      const newPathname = translateURL(pathname, locale, cookieLocale);
      // Just to be absolutely paranoidly certain it's not going to redirect
      // to the URL you're already don, we're doing this extra check.
      if (newPathname !== pathname) {
        navigate(newPathname);
      }
    }

    // Remember that the locale was overriden even if the hash changes,
    // otherwise it breaks self-links.
    // Remove the hash so the url looks cleaner.
    if (hasLocaleOverrideHash) {
      setPreventLocaleOverride(true);
      window.history.pushState(
        { resetLocaleOverride: true },
        "",
        window.location.href.replace(/#.*$/, "")
      );
    }
  }, [preventLocaleOverride, locale, hash, pathname, navigate, translations]);
```

Reply:
This should fix the self-links, and preserve forward/back navigation.